### PR TITLE
imba benchmark

### PR DESCRIPTION
**Pull request by allain**

### CODE REVIEW

File: frameworks/keyed/imba/src/main.imba [hunk 1]
```diff
@@ -0,0 +1,120 @@
+var A = ["pretty", "large", "big", "small", "tall", "short", "long", "handsome", "plain", "quaint", "clean", "elegant", "easy", "angry", "crazy", "helpful", "mushy", "odd", "unsightly", "adorable", "important", "inexpensive", "cheap", "expensive", "fancy"]
+var C = ["red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black", "orange"]
+var N = ["table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger", "pizza", "mouse",  "keyboard"]
+
+tag GlyphIcon < span
```
Comment (localvoid) [inline frameworks/keyed/imba/src/main.imba]: It is better to inline `GlyphIcon` into the `Row`, Imba should automatically detect that it is a static node.

Comment (allain) [reply frameworks/keyed/imba/src/main.imba]: Thanks. Done.

**Comment by krausest** [comment]
Sorry - I tried to merge, but ran into an issue:
For some reason I don't understand the buttons don't get id attributes which means the test driver can't run the benchmark:
bildschirmfoto 2018-12-04 um 21 05 51
Can you please take a look at it?

**Comment by leeoniya** [comment]
maybe this will work?

```
<button.btn.btn-primary.btn-block :type='button' :id=@id
```

append `#@id`:

```
<button.btn.btn-primary.btn-block#@id :type='button' :id=@id
```

**Comment by krausest** [comment]
Nice try, but...
```
ERROR in ./src/main.imba
Module Error (from ./node_modules/imba/loader.js):
Parse error at [969:3]: Unexpected 'IDENTIFIER'
      30 |              <self>
      31 |                      <div.col-sm-6.smallpad>
   -> 32 |                              <button.btn.btn-primary.btn-block#@id :type='button' :id=@id :tap=@cb> @title
```
(stays the same when removing the :id or pushing @id to the front)

**Comment by allain** [comment]
I'll take a stab shortly.

**Comment by allain** [comment]
I'm not sure if this particular test belongs in keyed or unkeyed to be honest.

**Comment by krausest** [comment]
Thanks - I‘ll take a look at it tomorrow.

**Comment by leeoniya** [comment]
> I'm not sure if this particular test belongs in keyed or unkeyed to be honest.

`npm run isKeyed -- --framework svelte` should tell you

**Comment by allain** [comment]
I mean I know wrote it in the keyed directory but I'm not sure if the code
satisfies the definition of a keyed framework test.

**Comment by localvoid** [comment]
> @allain I'm not sure if the code satisfies the definition of a keyed framework test.

I think that keys are assigned with `@{key}` syntax: https://github.com/somebee/dom-reconciler-bench/blob/29041f8cd28a4e38f9a830ad8ab3cc2adbe44550/apps/imba/app.imba#L86

**Comment by allain** [comment]
@localvoid You are absolutely correct, I've updated the Rows to be rendered with a key according to the syntax you pointed to.

Thank you.

I'll copy that one over to the unkeyed directory and remove the key syntax too.

**Comment by localvoid** [comment]
@allain 

```sh
$ cd webdriver-ts
$ npm run isKeyed -- --framework imba-v1.4.1-keyed

> webdriver-ts@1.0.0 isKeyed /home/void/w/js/garbage/js-framework-benchmark/webdriver-ts
> cross-env LANG="en_US.UTF-8" node dist/isKeyed.js "--framework" "imba-v1.4.1-keyed"

Frameworks that will be checked imba-v1.4.1-keyed
imba-v1.4.1-keyed is non-keyed for 'run benchmark' and non-keyed for 'remove row benchmark' non-keyed for 'swap rows benchmark' . It'll appear as non-keyed in the results
ERROR: Framework imba-v1.4.1-keyed is not correctly categorized
```

Key should be assigned on the `<Row>`:

```
<Row@{item:id} item=item selected=(selected === item:id) select=(do select(item)) remove=(do remove(item))> 
```

```sh
$ npm run isKeyed -- --framework imba-v1.4.1-keyed

> webdriver-ts@1.0.0 isKeyed /home/void/w/js/garbage/js-framework-benchmark/webdriver-ts
> cross-env LANG="en_US.UTF-8" node dist/isKeyed.js "--framework" "imba-v1.4.1-keyed"

Frameworks that will be checked imba-v1.4.1-keyed
imba-v1.4.1-keyed is keyed for 'run benchmark' and keyed for 'remove row benchmark' keyed for 'swap rows benchmark' . It'll appear as keyed in the results
```

File: frameworks/keyed/imba/src/main.imba [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	prop selected
+
+	def onSelect
+		@select(@item)
+
+	def onRemove 
+		@remove(@item)
+
+	def render
+		<self@{item:id} .selected=@selected>
```
Comment (localvoid) [inline frameworks/keyed/imba/src/main.imba]: Should be `<self .danger=@selected>`

File: frameworks/keyed/imba/src/main.imba [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			<td.col-md-1> @item:id
+			<td.col-md-4><a :tap.onSelect> @item:label
+			<td.col-md-1><a :tap.onRemove> <span.glyphicon.glyphicon-remove :aria-hidden=true>
```
Comment (localvoid) [inline frameworks/keyed/imba/src/main.imba]: When I checked in dev tools, `glyphicon` doesn't have `aria-hidden` attribute, so I guess it should be `<td.col-md-1><a :tap.onRemove> <span.glyphicon.glyphicon-remove aria-hidden=true>`, but it seems that Imba doesn't support aria attributes and I got an error `Uncaught TypeError: _1(...).flag(...).flag(...).flag(...).setAriaHidden is not a function`

File: frameworks/keyed/imba/src/main.imba [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+tag Button
+	prop id
+	prop cb
+	prop title
+
+	def render
+		<self>
+			<div.col-sm-6.smallpad>
+				<button.btn.btn-primary.btn-block :type='button' id=@id :tap=@cb> @title
```
Comment (localvoid) [inline frameworks/keyed/imba/src/main.imba]: type should be assigned as DOM attribute `<button.btn.btn-primary.btn-block type='button' id=@id :tap=@cb> @title`

File: frameworks/keyed/imba/src/main.imba [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+								<Button id="runlots" title="Create 10,000 rows" cb=(do runLots)>
+								<Button id="add" title="Append 1,000 rows" cb=(do add)>
+								<Button id="update" title="Update every 10th row" cb=(do update)>
+								<Button id="clear" title="Clear" cb=(do clear)>
+								<Button id="swaprows" title="Swap Rows" cb=(do swapRows)>
+
+				<table.table.table-hover.table-striped.test-data>
+					<tbody> for item in items
+						<Row :key=item:id item=item selected=(selected === item:id) select=(do select(item)) remove=(do remove(item))> 
+				<span.preloadicon.glyphicon.glyphicon-remove :aria-hidden=true>
```
Comment (localvoid) [inline frameworks/keyed/imba/src/main.imba]: Again, `aria-hidden` should be assigned as DOM attribute.

**Comment by krausest** [comment]
@allain I can't merge it in the current state. Can you please update it to generate the correct html and the class for selected rows as @localvoid reviewed?
Since the non-keyed version works (after changing the class for selected rows) I took a first look at the performance (though it's not really comparable until the identical HTML is generated):
bildschirmfoto 2018-12-05 um 20 52 56

**Comment by localvoid** [comment]
As a temporary solution it is possible to assign `aria-hidden` this way:

```
tag GlyphIcon < span
	def build
		set('aria-hidden', true)

	def render
		<self.glyphicon.glyphicon-remove>
```

**Comment by allain** [comment]
Thanks. I'll go ahead and apply that change.

**Comment by allain** [comment]
Done. Thank you Boris

**Comment by krausest** [comment]
@allain Keyed looks perfect, thanks!
The non-keyed version is identical, isn't it? (At least it's keyed too according to my test)
I could merge the PR with just the keyed version if that's ok for you.

**Comment by krausest** [comment]
Non-keyed is fine after I removed it. I hope to post results tomorrow.

**Comment by allain** [comment]
Woohoo. Thanks!

**Comment by PierBover** [comment]
@krausest please let us know when we can see the results!

**Comment by krausest** [comment]
@allain @PierBover 
First results - please note that only imba ran on chrome 71, all other benchmarks were measured on chrome 70. Hopefully I can rerun all benchmarks this weekend.
bildschirmfoto 2018-12-07 um 19 50 15

**Comment by PierBover** [comment]
Maybe @somebee can comment on that

**Comment by krausest** [comment]
@allain @PierBover 
The benchmark has been updated for Chrome 71:
bildschirmfoto 2018-12-08 um 11 14 43

https://krausest.github.io/js-framework-benchmark/current.html

**Pull request merged**